### PR TITLE
Treat 'abstract' properties the same as 'virtual' for src-gen

### LINKED ISSUE #58985: JsonSourceGenerator - ArgumentException key already added.
### Description

6.0.100-preview.7.21379.14 on Windows x64

Following code crashes the JsonSourceGenerator:

```c#
    [JsonSerializable(typeof(B))]
    public partial class MySerializerContext : JsonSerializerContext
    {
    }

    public class B : A
    {
        [JsonIgnore]
        public override int I => 0;
    }

    public abstract class A
    {
        [JsonIgnore]
        public abstract int I { get; }
    }
```

With the following:
```
1>CSC : warning CS8785: Generator 'JsonSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'An item with the same key has already been added.'
```

Workarounds:
Interestingly if I change A.I to be virtual instead of abstract, it works.
Also, if I remove JsonIgnore from the base (and keep it abstract), it also works.
So the workarounds are easy.

```
 	System.Text.Json.SourceGeneration.dll!System.Text.Json.SourceGeneration.JsonSourceGenerator.Parser.CacheMember(System.Text.Json.SourceGeneration.PropertyGenerationSpec propGenSpec, ref System.Collections.Generic.List<System.Text.Json.SourceGeneration.PropertyGenerationSpec> propGenSpecList, ref System.Collections.Generic.Dictionary<string, System.Text.Json.SourceGeneration.PropertyGenerationSpec> ignoredMembers)	Unknown
 	System.Text.Json.SourceGeneration.dll!System.Text.Json.SourceGeneration.JsonSourceGenerator.Parser.GetOrAddTypeGenerationSpec(System.Type type, System.Text.Json.Serialization.JsonSourceGenerationMode generationMode)	Unknown
 	System.Text.Json.SourceGeneration.dll!System.Text.Json.SourceGeneration.JsonSourceGenerator.Parser.GetRootSerializableType(Microsoft.CodeAnalysis.SemanticModel compilationSemanticModel, Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax attributeSyntax, System.Text.Json.Serialization.JsonSourceGenerationMode generationMode)	Unknown
 	System.Text.Json.SourceGeneration.dll!System.Text.Json.SourceGeneration.JsonSourceGenerator.Parser.GetGenerationSpec(System.Collections.Generic.List<Microsoft.CodeAnalysis.CSharp.Syntax.ClassDeclarationSyntax> classDeclarationSyntaxList)	Unknown
 	System.Text.Json.SourceGeneration.dll!System.Text.Json.SourceGeneration.JsonSourceGenerator.Execute(Microsoft.CodeAnalysis.GeneratorExecutionContext executionContext)	Unknown
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.GeneratorDriver.RunGeneratorsCore(Microsoft.CodeAnalysis.Compilation compilation, Microsoft.CodeAnalysis.DiagnosticBag diagnosticsBag, System.Threading.CancellationToken cancellationToken)	Unknown
```

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/58985 for 7.0.

Expected to be ported to 6.0.

### CODE REVIEW

File: src/libraries/System.Text.Json/Common/ReflectionExtensions.cs [hunk 1]
```diff
@@ -228,7 +228,14 @@ private static string GetBaseNameFromGenericTypeDef(Type genericTypeDef)
...
         {
             Debug.Assert(propertyInfo != null);
-            return propertyInfo != null && (propertyInfo.GetMethod?.IsVirtual == true || propertyInfo.SetMethod?.IsVirtual == true);
+            return propertyInfo != null &&
+                (propertyInfo.GetMethod?.IsVirtual == true || propertyInfo.SetMethod?.IsVirtual == true
+#if BUILDING_SOURCE_GENERATOR
+                // The Roslyn PropertyInfo.IsVirtual extension is based on ISymbol and does not match
+                // reflection semantics which includes IsAbstract, so add that here.
+                || propertyInfo.GetMethod?.IsAbstract == true || propertyInfo.SetMethod?.IsAbstract == true
+#endif
```
Comment (layomia) [inline src/libraries/System.Text.Json/Common/ReflectionExtensions.cs:237]: Would an alternative here be to add `attributes |= MethodAttributes.Virtual;` on L41 here?

https://github.com/dotnet/runtime/blob/a8eaaa7473d462fdccccf38504bfc7ad803347b8/src/libraries/System.Text.Json/gen/Reflection/RoslynExtensions.cs#L40

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs [hunk 2]
```diff
@@ -303,6 +303,61 @@ public virtual void RoundTripWithCustomPropertyConverterFactory_Struct()
             }
         }
 
+        [Fact]
+        public virtual void RoundTripConcreteDerivedClass()
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs:307]: I recommend sharing this test with the dynamic serializer in this file instead - https://github.com/dotnet/runtime/blob/7a58bef7497f70ab6bb98efc5f0b3f6b4ea2f4f2/src/libraries/System.Text.Json/tests/Common/PropertyVisibilityTests.cs. There are no `abstract` property tests there right now.

File: src/libraries/System.Text.Json/Common/ReflectionExtensions.cs (hunk 1 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/Common/ReflectionExtensions.cs]: Thanks. For some reason, I thought those extensions were actually in Roslyn code, not STJ.

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs (hunk 2 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs]: Sure thanks.

**Review by layomia — APPROVED** [review]
LGTM

**Review by eiriktsarpalis — APPROVED** [review]

**Comment by steveharter** [comment]
Build timeouts not related to this PR.

Test failure in collections see https://github.com/dotnet/runtime/issues/58933

**Pull request merged**